File: ui/react-auth-skeleton/src/services/shipmntsService.js
import apiClient from './apiClient';

let inFlightRequest = null;

export const shipmntsService = {
  async fetchData() {
    if (!inFlightRequest) {
      inFlightRequest = apiClient.get('/shipmnts/fetch')
        .then((response) => response.data?.data ?? response.data)
        .finally(() => { inFlightRequest = null; });
    }
    return inFlightRequest;
  },

  async generateRefreshToken({ email, password, interactive } = {}) {
    const body = {};
    if (email) body.email = email;
    if (password) body.password = password;
    if (interactive) body.interactive = true;
    const res = await apiClient.post('/shipmnts/authenticate', body);
    return res.data?.data || {};
  },

  async fetchProfile() {
    const res = await apiClient.get('/shipmnts/fetch');
    return res.data?.data || {};
  },

  async getAuthorizationUrl() {
    const res = await apiClient.get('/shipmnts/oauth/authorize-url');
    return res.data?.data?.authorizeUrl || '';
  },

  async exchangeAuthorizationCode(code) {
    const res = await apiClient.post('/shipmnts/oauth/callback', { code });
    return res.data?.data || {};
  },
};

export default shipmntsService;

File: identity/src/main/java/com/spider/sevice/integration/ShipmntsService.java
package com.spider.sevice.integration;

import com.fasterxml.jackson.databind.JsonNode;

public interface ShipmntsService {
    JsonNode fetchData();

    /**
     * Triggers the authentication flow to generate a new refresh token.
     * This will execute the npm script that opens a browser for Shipmnts login.
     * If email and password are provided they will be used for the authentication flow.
     * Returns the new refresh token (may be blank if not returned by the script).
     * If interactive is true the script will run with a visible browser so user can complete CAPTCHA/MFA.
     */
    String generateRefreshToken(String email, String password, boolean interactive);

    /**
     * Exchanges an OAuth authorization code for access and refresh tokens.
     * This implements the OAuth 2.0 authorization code flow.
     * Returns the refresh token obtained from the exchange.
     */
    String exchangeAuthorizationCode(String authorizationCode);

    /**
     * Gets the OAuth authorization URL for initiating the OAuth flow.
     * The user's browser should navigate to this URL.
     */
    String getAuthorizationUrl();
}
